[WIP]  Add abortTest() helper function

gernest opened the pull request:
Closes #1001
This adds abortTest() helper function to the k6 module. This function when
called inside a script it will

- stop the whole test run and k6 will exit with 107 status code
- stops immediately the VU that called it  and no more iterations are started
- make sure the teardown() is called
- the engine run status is 7 (RunStatusAbortedScriptError)

`(*goja.Runtime).Interrupt` is used for halting script execution and capturing
stack traces for better error message of what is happening with the script.

We introduce InterruptError which is used with `(*goja.Runtime).Interrupt` to
identify interrupts emitted by abortTest(). This way we use special handling of
this type.

Example script is

```js
import { abortTest, sleep } from 'k6';

export default function () {
    // We abort the test on second iteration
    if (__ITER == 1) {
        abortTest();
    }
    sleep(1);
}

export function teardown() {
    console.log('This function will be called even when we abort script');
}
```

abortTest() can be called in both default and setup functions, however you can't use it
in the init context

The following script will fail with the error
```
ERRO[0000] Using abortTest() in the init context is not supported at (...path to the script )init.js:13:43(34)
```

```js
import {
    abortTest
} from 'k6';

abortTest();

export function setup() {
}

export default function () {
    // ... some test logic ...
    console.log('mayday, mayday');
}
```

You can customize the reason for abortTest() by passing values to the function

```js
abortTest("Exceeded expectations");
```

Will emit `"Exceeded expectations"` on the logs

Missing 
 - [ ] Tests
<!--


  (ﾉ◕ヮ◕)ﾉ*:・ﾟ✧
  
  Thank you for your interest in contributing to the k6 project!
  
  Before you get started, we'd kindly like to ask you to read our:
    - Contribution guidelines at https://github.com/loadimpact/k6/blob/master/CONTRIBUTING.md
    - Code of Conduct at https://github.com/loadimpact/k6/blob/master/CODE_OF_CONDUCT.md
    
  Out of respect for your time, please start a discussion regarding any bigger contributions either
  in a GitHub Issue, in the community forums or in the #contributors channel of the k6 slack before you
  get started on the implementation.
  
  If you've already done all of that, you're more than welcome to proceed with your pull request.
  Thank you again for your contribution! 🙏🏼
  
  
-->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/loadimpact/k6?pullRequest=1920) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/loadimpact/k6?pullRequest=1920) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/loadimpact/k6?pullRequest=1920) it.</sub>

na-- reviewed (changes requested):
Thanks! I haven't done a full code review yet, I've only skimmed the changes and played around with the resulting binary a bit, but in general, your approach looks good! The only big issue I noticed is that these changes break the  `--linger` functionality. 

`k6 run --linger script.js` should still linger when `abortTest()` is used, i.e. the k6 _process_ shouldn't exit directly after the test run has stopped. This behavior should remain the same as it was with the old way of aborting the test run (thresholds with `abortOnFail: true`).

gernest commented:
@na-- thanks for the quick and good feedback. I had missed the `--linger` flag, but the docs were informative so I was able to address the flag use case.

When we run scripts that use `abortTest()` with  `--linger`  ,  basically every thing described on the PR message above stays the same except we don't immediately exit with `107` code .

Winding down of the engine  will be observed like with thresholds and waiting for `CTRL+C` which will  then allow us to exit with `107`.

Relevant commit c07af18

imiric commented on lib/executor/helpers.go:
```diff
@@ -74,6 +75,60 @@ func validateStages(stages []Stage) []error {
+
+// CancelExecutorContext cancels executor context found in ctx, ctx can be a
+// child of a context that was created with Context function.
+func CancelExecutorContext(ctx context.Context, err error) {
```
This can be unexported, right?

gernest replied:
Yep! initially the cancellable context was initialised independently in each executor, that is no longer the case though.

done e58e3eb5

MStoykov commented on js/modules/k6/k6.go:
```diff
@@ -116,6 +116,26 @@ func (*K6) Group(ctx context.Context, name string, fn goja.Callable) (goja.Value
+// AbortTest exposes abortTest function in the k6 module. When called ti will
+// interrupt the active goja runtime passed with ctx.
+func (*K6) AbortTest(ctx context.Context, extra ...goja.Value) {
+	state := lib.GetState(ctx)
+	rt := common.GetRuntime(ctx)
+	if state == nil {
+		rt.Interrupt(&common.InterruptError{Reason: common.AbortTestInitContext})
+		return
+	}
```
as mentioned above, I think you should just drop this lines

gernest replied:
done eaa9ea97

MStoykov commented on lib/executor/helpers.go:
```diff
@@ -28,6 +28,7 @@ import (
 
 	"github.com/sirupsen/logrus"
 
+	"github.com/loadimpact/k6/js/common"
```
you should not be importing `k6/js/*` anything from anywhere in `lib`. 

This has many reasons but in general we don't want to tie the `lib` functionality to a specific runner even if we only currently have `js` one and it is unlikely we will ever support others(and there are many other problems in the way of it)

gernest replied:
I moved `InterruptError` to `lib/errors`. 

done 75f8e87b

MStoykov commented on lib/executor/helpers.go:
```diff
@@ -96,6 +151,10 @@ func getIterationRunner(
 			return false
 		default:
 			if err != nil {
+				if handleInterupt(ctx, err) {
+					executionState.AddInterruptedIterations(1)
+					return false
+				}
```
Looking at the code nothing uses the return value. This seems to not matter as the contexts are getting checked constantly, but it is probably a good idea to figure out if we need to use it somehow

cc @imiric @na--

na-- replied:
Yeah, we might make a small PR to refactor this, but it might be needed eventually (e.g. in https://github.com/k6io/k6/issues/1250 or https://github.com/k6io/k6/issues/877) and is so small that it's probably fine as it is.

imiric reviewed (commented):
This looks very good to me, nice work! I just left some minor comments and nitpicks.

Also there's a check next to "Check teardown is called when a script aborts", but I don't see that test and I agree that it should be added.

MStoykov reviewed (changes requested):
Seems good in general :tada: 

- definitely, need to move the IterruptedError definition out of `js/common` and probably ... `lib/errors`, something else ?
- not so certain that this isn't an abuse of contexts and that I wouldn't prefer a more direct approach, but looking at the code around this ... I think that will require some pretty big refactoring and probably will make things actually worse than better, so I am fine with it

gernest commented on cmd/run.go:
```diff
@@ -276,8 +279,16 @@ a commandline interface for interacting with it.`,
+			var interrupt error
 			if err := engineRun(); err != nil {
-				return getExitCodeFromEngine(err)
+				if !conf.Linger.Bool {
+					return getExitCodeFromEngine(err)
+				}
+				if !common.IsInteruptError(err) {
+					return getExitCodeFromEngine(err)
+				}
```
thanks , nice catch

gernest commented on js/modules/k6/k6.go:
```diff
@@ -116,6 +116,26 @@ func (*K6) Group(ctx context.Context, name string, fn goja.Callable) (goja.Value
 
+// AbortTest exposes abortTest function in the k6 module. When called ti will
+// interrupt the active goja runtime passed with ctx.
+func (*K6) AbortTest(ctx context.Context, extra ...goja.Value) {
```
done 48a0ed6f

gernest commented on js/common/interupt_error.go:
```diff
@@ -0,0 +1,26 @@
+package common
```
done a7d66f94

imiric replied:
Nit: maybe make it `(C) 2021` :wink:

(I wish the `goheader` linter would work properly and do this for us...)

gernest replied:
done 921f35a1

gernest commented on js/common/interupt_error.go:
```diff
@@ -0,0 +1,26 @@
+// AbortTestInitContext is a reason  emitted when a test script calls abortTest() without arguments
+// in the init context
+const AbortTestInitContext = "Using abortTest() in the init context is not supported"
+
```
nice catch! 🤦  lol!

done eaa9ea97

gernest commented on cmd/run.go:
```diff
@@ -63,11 +64,13 @@ const (
 // TODO: fix this, global variables are not very testable...
 //nolint:gochecknoglobals
 var runType = os.Getenv("K6_TYPE")
+var errScriptInterupted = errors.New("script execution  was aborted")
```
done f2c9911fbce0edc3dfac41bbcb6a24c0a111bdc7

gernest commented:
@na-- @imiric @MStoykov thanks a lot for the amazing feedback, I have addressed all review comments (I think) PTAL

> Also there's a check next to "Check teardown is called when a script aborts", but I don't see that test and I agree that it should be added.

Thanks for catching this one, I added the test in b5b8441

na-- reviewed (approved):
LGTM! The only minor nitpick I have is that instead of directly type casting errors, you should probably use the recent [`errors.Is()`](https://golang.org/pkg/errors/#Is) and [`errors.As()`](https://golang.org/pkg/errors/#As) stdlib functions. 

Other than that, everything looks great! The changes ended up more complex than I expected, but can't think of a way to simplify them without significant refactoring and everything is still understandable, so this is going to be it! :tada:

imiric reviewed (approved):
LGTM, well done! Just left a tiny nitpick, feel free to disregard.

Also, please rebase on top of the latest `master` and cleanup the commits a bit before this is merged. Not necessarily squashing to a single commit, but whatever makes sense for the changes to be atomic.

rushby commented:
Hi all,

Just been looking into this functionality would be great if this can get merged in soon!

na-- commented:
@rushby, we wanted to merge this for k6 v0.33.0, but we weren't able because it's blocked by some other changes we needed to make in k6 (e.g. https://github.com/k6io/k6/pull/1863) and in our cloud, so it can work both in single instance and in distributed tests. We plan to merge it for k6 v0.34.0, slated to be released sometime in the end of August/early September.

imiric commented:
Closing this in favor of #2093.

We changed the API slightly but you'll get full credits @gernest. Thanks!
